File: scripts/parity-runners/python-scenarios/tokenize.py
from common import *  # noqa: F403
from common import _token_stream_next

def tokenize_token_stream(input_data: Any) -> dict[str, Any]:
    module = load_reference_token_stream()
    mode = input_data.get("mode", "closed_lifecycle")
    if not isinstance(mode, str):
        raise ValueError("mode must be a string")

    def stream_result(events: list[dict[str, Any]]) -> dict[str, Any]:
        return {
            "contract": "token-stream-" + mode.replace("_", "-"),
            "events": events,
        }

    def fields(text: str) -> list[str]:
        return text.split()

    if mode == "closed_lifecycle":
        stream = module.BufferedTokenStream(
            tokenize_fnc=fields,
            min_token_len=1,
            min_ctx_len=1,
        )
        before = stream._event_ch.closed
        error = False
        try:
            stream.end_input()
        except Exception:
            error = True
        after = stream._event_ch.closed
        return stream_result(
            [
                {"name": "closed_before", "result": before},
                {"name": "close", "error": error, "error_class": "error" if error else ""},
                {"name": "closed_after", "result": after},
            ]
        )

    if mode == "close_flush":
        stream = module.BufferedTokenStream(
            tokenize_fnc=lambda text: [text],
            min_token_len=1,
            min_ctx_len=1,
        )
        push_error = False
        close_error = False
        try:
            stream.push_text("hello")
        except Exception:
            push_error = True
        try:
            stream.end_input()
        except Exception:
            close_error = True
        return stream_result(
            [
                {
                    "name": "push_text",
                    "error": push_error,
                    "error_class": "error" if push_error else "",
                },
                {
                    "name": "close",
                    "error": close_error,
                    "error_class": "error" if close_error else "",
                },
                asyncio.run(_token_stream_next(stream)),
            ]
        )

    if mode == "next_eof_closed":
        stream = module.BufferedTokenStream(
            tokenize_fnc=lambda text: [],
            min_token_len=1,
            min_ctx_len=1,
        )
        close_error = False
        try:
            stream.end_input()
        except Exception:
            close_error = True
        return stream_result(
            [
                {
                    "name": "close",
                    "error": close_error,
                    "error_class": "error" if close_error else "",
                },
                asyncio.run(_token_stream_next(stream)),
            ]
        )

    if mode == "last_token_context":
        stream = module.BufferedTokenStream(
            tokenize_fnc=fields,
            min_token_len=1,
            min_ctx_len=1,
        )
        push_error = False
        flush_error = False
        try:
            stream.push_text("one two three")
        except Exception:
            push_error = True
        first = asyncio.run(_token_stream_next(stream))
        second = asyncio.run(_token_stream_next(stream))
        try:
            stream.flush()
        except Exception:
            flush_error = True
        third = asyncio.run(_token_stream_next(stream))
        return stream_result(
            [
                {
                    "name": "push_text",
                    "error": push_error,
                    "error_class": "error" if push_error else "",
                },
                first,
                second,
                {"name": "flush", "error": flush_error, "error_class": "error" if flush_error else ""},
                third,
            ]
        )

    if mode == "end_input_flush_close":
        stream = module.BufferedTokenStream(
            tokenize_fnc=lambda text: [text],
            min_token_len=1,
            min_ctx_len=10,
        )
        push_error = False
        end_error = False
        try:
            stream.push_text("hello")
        except Exception:
            push_error = True
        try:
            stream.end_input()
        except Exception:
            end_error = True
        return stream_result(
            [
                {
                    "name": "push_text",
                    "error": push_error,
                    "error_class": "error" if push_error else "",
                },
                {
                    "name": "end_input",
                    "error": end_error,
                    "error_class": "error" if end_error else "",
                },
                asyncio.run(_token_stream_next(stream)),
                asyncio.run(_token_stream_next(stream)),
            ]
        )

    if mode == "end_input_closed":
        stream = module.BufferedTokenStream(
            tokenize_fnc=fields,
            min_token_len=1,
            min_ctx_len=1,
        )
        first_error = False
        second_error = False
        try:
            stream.end_input()
        except Exception:
            first_error = True
        try:
            stream.end_input()
        except RuntimeError:
            second_error = True
        return stream_result(
            [
                {
                    "name": "first_end_input",
                    "error": first_error,
                    "error_class": "error" if first_error else "",
                },
                {
                    "name": "second_end_input",
                    "error": second_error,
                    "error_class": "error" if second_error else "",
                },
            ]
        )

    if mode == "aclose_no_flush":
        stream = module.BufferedTokenStream(
            tokenize_fnc=lambda text: [text],
            min_token_len=1,
            min_ctx_len=10,
        )
        push_error = False
        close_error = False
        try:
            stream.push_text("hello")
        except Exception:
            push_error = True
        try:
            asyncio.run(stream.aclose())
        except Exception:
            close_error = True
        return stream_result(
            [
                {
                    "name": "push_text",
                    "error": push_error,
                    "error_class": "error" if push_error else "",
                },
                {
                    "name": "aclose",
                    "error": close_error,
                    "error_class": "error" if close_error else "",
                },
                asyncio.run(_token_stream_next(stream)),
            ]
        )

    if mode == "whitespace_context":
        def whitespace_tokenizer(text: str) -> list[str]:
            if text.startswith("\t"):
                return ["\t", "two"]
            return text.split()

        stream = module.BufferedTokenStream(
            tokenize_fnc=whitespace_tokenizer,
            min_token_len=1,
            min_ctx_len=1,
        )
        push_error = False
        flush_error = False
        try:
            stream.push_text("one\t two")
        except Exception:
            push_error = True
        try:
            stream.flush()
        except Exception:
            flush_error = True
        return stream_result(
            [
                {
                    "name": "push_text",
                    "error": push_error,
                    "error_class": "error" if push_error else "",
                },
                {"name": "flush", "error": flush_error, "error_class": "error" if flush_error else ""},
                asyncio.run(_token_stream_next(stream)),
                asyncio.run(_token_stream_next(stream)),
            ]
        )

    raise ValueError(f"unknown token stream mode {mode}")


def tokenize_replace_words(input_data: Any) -> dict[str, Any]:
    return load_python_utils_runner().run_tokenize_replace_words(input_data)


def tokenize_format_words(input_data: Any) -> dict[str, Any]:
    values = input_data.get("word_values", [["hello", "world"]])
    if not isinstance(values, list) or not all(
        isinstance(words, list) and all(isinstance(word, str) for word in words)
        for words in values
    ):
        raise ValueError("word_values must be a list of string lists")
    events = [
        {"name": "format_words", "input": words, "result": " ".join(words)}
        for words in values
    ]
    return {"contract": "tokenize-format-words", "events": events}


def tokenize_sentence_tokenizer(input_data: Any) -> dict[str, Any]:
    runner = load_python_utils_runner()
    basic_sent = runner.load_reference_tokenize_module("_basic_sent", "_basic_sent.py")
    values = runner.string_values(
        input_data,
        "text_values",
        ["Version 1.5 is ready. Next sentence."],
    )
    events = []
    for value in values:
        result = [
            token
            for token, _start, _end in basic_sent.split_sentences(
                value,
                min_sentence_len=20,
                retain_format=False,
            )
        ]
        events.append({"name": "sentence_tokenize", "input": value, "result": result})
    return {"contract": "tokenize-sentence-tokenizer", "events": events}


def tokenize_split_words(input_data: Any) -> dict[str, Any]:
    return load_python_utils_runner().run_tokenize_split_words(input_data)


def tokenize_split_sentences(input_data: Any) -> dict[str, Any]:
    return load_python_utils_runner().run_tokenize_split_sentences(input_data)


def tokenize_paragraphs(input_data: Any) -> dict[str, Any]:
    runner = load_python_utils_runner()
    basic_paragraph = runner.load_reference_tokenize_module(
        "_basic_paragraph", "_basic_paragraph.py"
    )
    values = runner.string_values(input_data, "text_values", [" one\n\n two\nthree "])
    with_offset = bool(input_data.get("with_offset", False))
    events = []
    for value in values:
        paragraphs = basic_paragraph.split_paragraphs(value)
        if with_offset:
            result = [
                {"token": token, "start": start, "end": end}
                for token, start, end in paragraphs
            ]
            events.append({"name": "split_paragraphs", "input": value, "result": result})
        else:
            events.append(
                {
                    "name": "tokenize_paragraphs",
                    "input": value,
                    "result": [token for token, _start, _end in paragraphs],
                }
            )
    return {"contract": "tokenize-paragraphs", "events": events}


def tokenize_hyphenate_words(input_data: Any) -> dict[str, Any]:
    runner = load_python_utils_runner()
    basic_hyphenator = runner.load_reference_tokenize_module(
        "_basic_hyphenator", "_basic_hyphenator.py"
    )
    values = runner.string_values(
        input_data,
        "word_values",
        ["beautiful", "communication", "word"],
    )
    hyphenator = basic_hyphenator.Hyphenator(
        basic_hyphenator.PATTERNS,
        basic_hyphenator.EXCEPTIONS,
    )
    events = [
        {
            "name": "hyphenate_word",
            "input": value,
            "result": hyphenator.hyphenate_word(value),
        }
        for value in values
    ]
    return {"contract": "tokenize-hyphenate-words", "events": events}

File: scripts/parity-runners/go-scenario/tokenize.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"

	"github.com/cavos-io/rtp-agent/library/tokenize"
)

func runTokenizeTokenStream(input json.RawMessage) (any, error) {
	var payload struct {
		Mode string `json:"mode"`
	}
	if err := json.Unmarshal(input, &payload); err != nil {
		return nil, err
	}
	mode := payload.Mode
	if mode == "" {
		mode = "closed_lifecycle"
	}

	switch mode {
	case "closed_lifecycle":
		stream := tokenize.NewBufferedTokenStream(strings.Fields, 1, 1)
		before := stream.Closed()
		closeErr := stream.Close()
		after := stream.Closed()
		return tokenStreamResult(mode, []map[string]any{
			{"name": "closed_before", "result": before},
			{"name": "close", "error": closeErr != nil, "error_class": errorClass(closeErr)},
			{"name": "closed_after", "result": after},
		}), nil
	case "close_flush":
		stream := tokenize.NewBufferedTokenStream(func(text string) []string { return []string{text} }, 1, 1)
		pushErr := stream.PushText("hello")
		closeErr := stream.Close()
		token, nextErr := stream.Next()
		return tokenStreamResult(mode, []map[string]any{
			{"name": "push_text", "error": pushErr != nil, "error_class": errorClass(pushErr)},
			{"name": "close", "error": closeErr != nil, "error_class": errorClass(closeErr)},
			tokenStreamNextEvent("next", token, nextErr),
		}), nil
	case "next_eof_closed":
		stream := tokenize.NewBufferedTokenStream(func(string) []string { return nil }, 1, 1)
		closeErr := stream.Close()
		token, nextErr := stream.Next()
		return tokenStreamResult(mode, []map[string]any{
			{"name": "close", "error": closeErr != nil, "error_class": errorClass(closeErr)},
			tokenStreamNextEvent("next", token, nextErr),
		}), nil
	case "last_token_context":
		stream := tokenize.NewBufferedTokenStream(strings.Fields, 1, 1)
		pushErr := stream.PushText("one two three")
		first, firstErr := stream.Next()
		second, secondErr := stream.Next()
		flushErr := stream.Flush()
		third, thirdErr := stream.Next()
		return tokenStreamResult(mode, []map[string]any{
			{"name": "push_text", "error": pushErr != nil, "error_class": errorClass(pushErr)},
			tokenStreamNextEvent("next", first, firstErr),
			tokenStreamNextEvent("next", second, secondErr),
			{"name": "flush", "error": flushErr != nil, "error_class": errorClass(flushErr)},
			tokenStreamNextEvent("next", third, thirdErr),
		}), nil
	case "end_input_flush_close":
		stream := tokenize.NewBufferedTokenStream(func(text string) []string { return []string{text} }, 1, 10)
		pushErr := stream.PushText("hello")
		endErr := stream.EndInput()
		first, firstErr := stream.Next()
		second, secondErr := stream.Next()
		return tokenStreamResult(mode, []map[string]any{
			{"name": "push_text", "error": pushErr != nil, "error_class": errorClass(pushErr)},
			{"name": "end_input", "error": endErr != nil, "error_class": errorClass(endErr)},
			tokenStreamNextEvent("next", first, firstErr),
			tokenStreamNextEvent("next", second, secondErr),
		}), nil
	case "end_input_closed":
		stream := tokenize.NewBufferedTokenStream(strings.Fields, 1, 1)
		firstErr := stream.EndInput()
		secondErr := stream.EndInput()
		return tokenStreamResult(mode, []map[string]any{
			{"name": "first_end_input", "error": firstErr != nil, "error_class": errorClass(firstErr)},
			{"name": "second_end_input", "error": secondErr != nil, "error_class": errorClass(secondErr)},
		}), nil
	case "aclose_no_flush":
		stream := tokenize.NewBufferedTokenStream(func(text string) []string { return []string{text} }, 1, 10)
		pushErr := stream.PushText("hello")
		closeErr := stream.AClose()
		token, nextErr := stream.Next()
		return tokenStreamResult(mode, []map[string]any{
			{"name": "push_text", "error": pushErr != nil, "error_class": errorClass(pushErr)},
			{"name": "aclose", "error": closeErr != nil, "error_class": errorClass(closeErr)},
			tokenStreamNextEvent("next", token, nextErr),
		}), nil
	case "whitespace_context":
		stream := tokenize.NewBufferedTokenStream(func(text string) []string {
			if strings.HasPrefix(text, "\t") {
				return []string{"\t", "two"}
			}
			return strings.Fields(text)
		}, 1, 1)
		pushErr := stream.PushText("one\t two")
		flushErr := stream.Flush()
		first, firstErr := stream.Next()
		second, secondErr := stream.Next()
		return tokenStreamResult(mode, []map[string]any{
			{"name": "push_text", "error": pushErr != nil, "error_class": errorClass(pushErr)},
			{"name": "flush", "error": flushErr != nil, "error_class": errorClass(flushErr)},
			tokenStreamNextEvent("next", first, firstErr),
			tokenStreamNextEvent("next", second, secondErr),
		}), nil
	default:
		return nil, fmt.Errorf("unknown token stream mode %q", mode)
	}
}

func tokenStreamResult(mode string, events []map[string]any) map[string]any {
	return map[string]any{
		"contract": "token-stream-" + strings.ReplaceAll(mode, "_", "-"),
		"events":   events,
	}
}

func tokenStreamNextEvent(name string, token *tokenize.TokenData, err error) map[string]any {
	event := map[string]any{
		"name":        name,
		"error":       err != nil,
		"error_class": tokenStreamErrorClass(err),
	}
	if token != nil {
		event["token"] = token.Token
	}
	return event
}

func tokenStreamErrorClass(err error) string {
	if err == nil {
		return ""
	}
	if errors.Is(err, io.EOF) {
		return "eof"
	}
	return "error"
}

func runTokenizeReplaceWords(input json.RawMessage) (any, error) {
	var payload struct {
		TextValues   []string          `json:"text_values"`
		Replacements map[string]string `json:"replacements"`
	}
	if err := json.Unmarshal(input, &payload); err != nil {
		return nil, err
	}
	if payload.TextValues == nil {
		payload.TextValues = []string{"Hello, WORLD! workflow stays.", "Do not replace flow inside workflow."}
	}
	if payload.Replacements == nil {
		payload.Replacements = map[string]string{"hello": "hi", "world": "there", "flow": "stream"}
	}

	events := make([]map[string]any, 0, len(payload.TextValues))
	for _, value := range payload.TextValues {
		events = append(events, map[string]any{
			"name":   "replace_words",
			"input":  value,
			"result": tokenize.ReplaceWords(value, payload.Replacements),
		})
	}
	return map[string]any{"contract": "tokenize-replace-words", "events": events}, nil
}

func runTokenizeFormatWords(input json.RawMessage) (any, error) {
	var payload struct {
		WordValues [][]string `json:"word_values"`
	}
	if err := json.Unmarshal(input, &payload); err != nil {
		return nil, err
	}
	if payload.WordValues == nil {
		payload.WordValues = [][]string{{"hello", "world"}}
	}
	tokenizer := tokenize.NewBasicWordTokenizer()
	events := make([]map[string]any, 0, len(payload.WordValues))
	for _, words := range payload.WordValues {
		events = append(events, map[string]any{
			"name":   "format_words",
			"input":  words,
			"result": tokenizer.FormatWords(words),
		})
	}
	return map[string]any{"contract": "tokenize-format-words", "events": events}, nil
}

func runTokenizeSentenceTokenizer(input json.RawMessage) (any, error) {
	var payload struct {
		TextValues []string `json:"text_values"`
	}
	if err := json.Unmarshal(input, &payload); err != nil {
		return nil, err
	}
	if payload.TextValues == nil {
		payload.TextValues = []string{"Version 1.5 is ready. Next sentence."}
	}
	tokenizer := tokenize.NewBasicSentenceTokenizer()
	events := make([]map[string]any, 0, len(payload.TextValues))
	for _, value := range payload.TextValues {
		events = append(events, map[string]any{
			"name":   "sentence_tokenize",
			"input":  value,
			"result": tokenizer.Tokenize(value, ""),
		})
	}
	return map[string]any{"contract": "tokenize-sentence-tokenizer", "events": events}, nil
}

func runTokenizeSplitWords(input json.RawMessage) (any, error) {
	var payload struct {
		IgnorePunctuation *bool    `json:"ignore_punctuation"`
		RetainFormat      *bool    `json:"retain_format"`
		SplitCharacter    *bool    `json:"split_character"`
		TextValues        []string `json:"text_values"`
	}
	if err := json.Unmarshal(input, &payload); err != nil {
		return nil, err
	}
	if payload.TextValues == nil {
		payload.TextValues = []string{" Hello, world!  keep-format? ", "alpha beta,gamma"}
	}
	ignorePunctuation := boolValue(payload.IgnorePunctuation, true)
	splitCharacter := boolValue(payload.SplitCharacter, false)
	retainFormat := boolValue(payload.RetainFormat, false)

	events := make([]map[string]any, 0, len(payload.TextValues))
	for _, value := range payload.TextValues {
		words := tokenize.SplitWords(value, ignorePunctuation, splitCharacter, retainFormat)
		result := make([]map[string]any, 0, len(words))
		for _, word := range words {
			result = append(result, map[string]any{
				"token": word.Token,
				"start": word.Start,
				"end":   word.End,
			})
		}
		events = append(events, map[string]any{
			"name":   "split_words",
			"input":  value,
			"result": result,
		})
	}
	return map[string]any{"contract": "tokenize-split-words", "events": events}, nil
}

func runTokenizeSplitSentences(input json.RawMessage) (any, error) {
	var payload struct {
		MinSentenceLen *int     `json:"min_sentence_len"`
		RetainFormat   *bool    `json:"retain_format"`
		TextValues     []string `json:"text_values"`
	}
	if err := json.Unmarshal(input, &payload); err != nil {
		return nil, err
	}
	if payload.TextValues == nil {
		payload.TextValues = []string{"Version 1.5 is ready. Next sentence.", "他说：“你好。” 下一句。"}
	}
	minSentenceLen := intValue(payload.MinSentenceLen, 20)
	retainFormat := boolValue(payload.RetainFormat, false)

	events := make([]map[string]any, 0, len(payload.TextValues))
	for _, value := range payload.TextValues {
		sentences := tokenize.SplitSentences(value, minSentenceLen, retainFormat)
		result := make([]string, 0, len(sentences))
		for _, sentence := range sentences {
			result = append(result, sentence.Token)
		}
		events = append(events, map[string]any{
			"name":   "split_sentences",
			"input":  value,
			"result": result,
		})
	}
	return map[string]any{"contract": "tokenize-split-sentences", "events": events}, nil
}

func runTokenizeParagraphs(input json.RawMessage) (any, error) {
	var payload struct {
		TextValues []string `json:"text_values"`
		WithOffset *bool    `json:"with_offset"`
	}
	if err := json.Unmarshal(input, &payload); err != nil {
		return nil, err
	}
	if payload.TextValues == nil {
		payload.TextValues = []string{" one\n\n two\nthree "}
	}
	withOffset := boolValue(payload.WithOffset, false)
	events := make([]map[string]any, 0, len(payload.TextValues))
	for _, value := range payload.TextValues {
		if withOffset {
			paragraphs := tokenize.SplitParagraphs(value)
			result := make([]map[string]any, 0, len(paragraphs))
			for _, paragraph := range paragraphs {
				result = append(result, map[string]any{
					"token": paragraph.Token,
					"start": paragraph.Start,
					"end":   paragraph.End,
				})
			}
			events = append(events, map[string]any{"name": "split_paragraphs", "input": value, "result": result})
			continue
		}
		events = append(events, map[string]any{
			"name":   "tokenize_paragraphs",
			"input":  value,
			"result": tokenize.TokenizeParagraphs(value),
		})
	}
	return map[string]any{"contract": "tokenize-paragraphs", "events": events}, nil
}

func runTokenizeHyphenateWords(input json.RawMessage) (any, error) {
	var payload struct {
		WordValues []string `json:"word_values"`
	}
	if err := json.Unmarshal(input, &payload); err != nil {
		return nil, err
	}
	if payload.WordValues == nil {
		payload.WordValues = []string{"beautiful", "communication", "word"}
	}
	events := make([]map[string]any, 0, len(payload.WordValues))
	for _, value := range payload.WordValues {
		events = append(events, map[string]any{
			"name":   "hyphenate_word",
			"input":  value,
			"result": tokenize.HyphenateWord(value),
		})
	}
	return map[string]any{"contract": "tokenize-hyphenate-words", "events": events}, nil
}
